src/lib/lightning/clightning/clightningService.ts

```diff
@@ -197,9 +199,100 @@ class CLightningService implements LightningService {
     );
   }
 
+  async addListenerToNode(node: LightningNode): Promise<void> {
+    console.log('addListenerToNode CLN on port: ', node.ports.rest);
+  }
+
+  private channelCaches: {
+    [nodePort: number]: {
+      intervalId: NodeJS.Timeout;
+      channels: { pending: boolean; status: string }[];
+    };
+  } = {};
+
+  async removeListener(node: LightningNode): Promise<void> {
+    const nodePort = this.getNodePort(node);
+    const cache = this.channelCaches[nodePort];
+    if (cache) {
+      clearInterval(cache.intervalId);
+      delete this.channelCaches[nodePort];
+    }
+  }
+
+  async subscribeChannelEvents(
+    node: LightningNode,
+    callback: (event: PLN.LightningNodeChannelEvent) => void,
+  ): Promise<void> {
+    const nodePort = this.getNodePort(node);
+    if (!this.channelCaches[nodePort]) {
+      // check c-lightning channels every 30 seconds
+      this.channelCaches[nodePort] = {
+        intervalId: setInterval(() => {
+          this.checkChannels(node, callback);
+        }, 30 * 1000),
+        channels: [],
+      };
+    }
+  }
+
+  checkChannels = async (
+    node: LightningNode,
+    callback: (event: PLN.LightningNodeChannelEvent) => void,
+  ) => {
+    const result = await httpGet<CLN.GetChannelsResponse[]>(node, 'channel/listChannels');
+    const channels = result.map(c => {
+      const status = ChannelStateToStatus[c.state];
+      return { pending: status !== 'Open', status };
+    });
+
+    const cache = this.channelCaches[this.getNodePort(node)];
+
+    if (!this.areChannelsEqual(cache.channels, channels)) {
+      this.handleUpdatedChannels(channels, callback);
+      cache.channels = channels;
+      // edge case for empty channels but cache channels is not empty
+      if (!channels.length) {
+        callback({ type: 'Closed' });
+      }
+    }
+  };
+
+  private handleUpdatedChannels = (
+    channels: { pending: boolean; status: string }[],
+    callback: (event: PLN.LightningNodeChannelEvent) => void,
+  ) => {
+    channels.forEach(channel => {
+      if (channel.pending) {
+        callback({ type: 'Pending' });
+      } else if (channel.status === 'Open') {
+        callback({ type: 'Open' });
+      } else if (channel.status === 'Closed') {
+        callback({ type: 'Closed' });
+      }
+    });
```

Comment:
This is going to call the `callback` multiple times. One for each channel, even if only one channel changed. The throttling will prevent syncing the chart more than once but from an API design perspective, this feels like unexpected behavior. The `subscribeChannelEvents` for other implementations only sends a message for the single channel that has been updated.

Reply:
Updated